### PR TITLE
[fix](index compaction)Fix MOW index compaction core

### LINKED ISSUE #32084: [Bug] index compaction core in MOW table with inverted index
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

branch-2.0
master 

### What's Wrong?

be core
```
start time: 2024年 03月 12日 星期二 10:24:52 CST
INFO: java_cmd /mnt/disk2/luen/develop/tools/jdk1.8.0_131/bin/java
INFO: jdk_version 8
doris_be: /mnt/disk2/luen/develop/workspace/doris/be/src/clucene/src/core/CLucene/index/IndexWriter.cpp:1755: void lucene::index::IndexWriter::mergeTerms(bool): Assertion `destDocId > lastDoc || df == 1' failed.
*** Query id: 0-0 ***
*** tablet id: 10163 ***
*** Aborted at 1710211055 (unix time) try "date -d @1710211055" if you are using GNU date ***
*** Current BE git commitID: 7d550a504f ***
*** SIGABRT unknown detail explain (@0x4210018f2d3) received by PID 1635027 (TID 1636865 OR 0x7f98545ce700) from PID 1635027; stack trace: ***
doris_be: /mnt/disk2/luen/develop/workspace/doris/be/src/clucene/src/core/CLucene/index/IndexWriter.cpp:1755: void lucene::index::IndexWriter::mergeTerms(bool): Assertion `destDocId > lastDoc || df == 1' failed.
doris_be: /mnt/disk2/luen/develop/workspace/doris/be/src/clucene/src/core/CLucene/index/IndexWriter.cpp:1755: void lucene::index::IndexWriter::mergeTerms(bool): Assertion `destDocId > lastDoc || df == 1' failed.
doris_be: /mnt/disk2/luen/develop/workspace/doris/be/src/clucene/src/core/CLucene/index/IndexWriter.cpp:1755: void lucene::index::IndexWriter::mergeTerms(bool): Assertion `destDocId > lastDoc || df == 1' failed.
doris_be: /mnt/disk2/luen/develop/workspace/doris/be/src/clucene/src/core/CLucene/index/IndexWriter.cpp:1755: void lucene::index::IndexWriter::mergeTerms(bool): Assertion `destDocId > lastDoc || df == 1' failed.
doris_be: /mnt/disk2/luen/develop/workspace/doris/be/src/clucene/src/core/CLucene/index/IndexWriter.cpp:1755: void lucene::index::IndexWriter::mergeTerms(bool): Assertion `destDocId > lastDoc || df == 1' failed.
 0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at /mnt/disk2/luen/develop/workspace/doris/be/src/common/signal_handler.h:417
 1# 0x00007F9BFE9F1B50 in /lib64/libc.so.6
 2# __GI_raise in /lib64/libc.so.6
 3# __GI_abort in /lib64/libc.so.6
 4# _nl_load_domain.cold.0 in /lib64/libc.so.6
 5# 0x00007F9BFE9EA426 in /lib64/libc.so.6
 6# lucene::index::IndexWriter::mergeTerms(bool) at /mnt/disk2/luen/develop/workspace/doris/be/src/clucene/src/core/CLucene/index/IndexWriter.cpp:1756
 7# lucene::index::IndexWriter::indexCompaction(std::vector<lucene::store::Directory*, std::allocator<lucene::store::Directory*> >&, std::vector<lucene::store::Directory*, std::allocator<lucene::store::Directory*> >, std::vector<std::vector<std::pair<unsigned int, unsigned int>, std::allocator<std::pair<unsigned int, unsigned int> > >, std::allocator<std::vector<std::pair<unsigned int, unsigned int>, std::allocator<std::pair<unsigned int, unsigned int> > > > >, std::vector<unsigned int, std::allocator<unsigned int> >) at /mnt/disk2/luen/develop/workspace/doris/be/src/clucene/src/core/CLucene/index/IndexWriter.cpp:1389
 8# doris::segment_v2::compact_column(int, int, int, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::shared_ptr<doris::io::FileSystem> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<std::vector<std::pair<unsigned int, unsigned int>, std::allocator<std::pair<unsigned int, unsigned int> > >, std::allocator<std::vector<std::pair<unsigned int, unsigned int>, std::allocator<std::pair<unsigned int, unsigned int> > > > >, std::vector<unsigned int, std::allocator<unsigned int> >) at /mnt/disk2/luen/develop/workspace/doris/be/src/olap/rowset/segment_v2/inverted_index_compaction.cpp:71
 9# doris::Compaction::do_compaction_impl(long)::$_1::operator()(int) const at /mnt/disk2/luen/develop/workspace/doris/be/src/olap/compaction.cpp:523
10# doris::Compaction::do_compaction_impl(long)::$_1 std::for_each<std::_Rb_tree_const_iterator<int>, doris::Compaction::do_compaction_impl(long)::$_1>(std::_Rb_tree_const_iterator<int>, std::_Rb_tree_const_iterator<int>, doris::Compaction::do_compaction_impl(long)::$_1) at /mnt/disk2/luen/develop/tools/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_algo.h:3819
11# doris::Compaction::do_compaction_impl(long) at /mnt/disk2/luen/develop/workspace/doris/be/src/olap/compaction.cpp:515
12# doris::Compaction::do_compaction(long) at /mnt/disk2/luen/develop/workspace/doris/be/src/olap/compaction.cpp:126
13# doris::CumulativeCompaction::execute_compact_impl() at /mnt/disk2/luen/develop/workspace/doris/be/src/olap/cumulative_compaction.cpp:79
14# doris::Compaction::execute_compact() at /mnt/disk2/luen/develop/workspace/doris/be/src/olap/compaction.cpp:108
15# doris::Tablet::execute_compaction(doris::Compaction&) at /mnt/disk2/luen/develop/workspace/doris/be/src/olap/tablet.cpp:1908
16# doris::StorageEngine::_submit_compaction_task(std::shared_ptr<doris::Tablet>, doris::CompactionType, bool)::$_0::operator()() const at /mnt/disk2/luen/develop/workspace/doris/be/src/olap/olap_server.cpp:969
17# void std::__invoke_impl<void, doris::StorageEngine::_submit_compaction_task(std::shared_ptr<doris::Tablet>, doris::CompactionType, bool)::$_0&>(std::__invoke_other, doris::StorageEngine::_submit_compaction_task(std::shared_ptr<doris::Tablet>, doris::CompactionType, bool)::$_0&) at /mnt/disk2/luen/develop/tools/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:61
18# std::enable_if<is_invocable_r_v<void, doris::StorageEngine::_submit_compaction_task(std::shared_ptr<doris::Tablet>, doris::CompactionType, bool)::$_0&>, void>::type std::__invoke_r<void, doris::StorageEngine::_submit_compaction_task(std::shared_ptr<doris::Tablet>, doris::CompactionType, bool)::$_0&>(doris::StorageEngine::_submit_compaction_task(std::shared_ptr<doris::Tablet>, doris::CompactionType, bool)::$_0&) at /mnt/disk2/luen/develop/tools/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:117
19# std::_Function_handler<void (), doris::StorageEngine::_submit_compaction_task(std::shared_ptr<doris::Tablet>, doris::CompactionType, bool)::$_0>::_M_invoke(std::_Any_data const&) at /mnt/disk2/luen/develop/tools/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:291
20# std::function<void ()>::operator()() const at /mnt/disk2/luen/develop/tools/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:560
21# doris::FunctionRunnable::run() at /mnt/disk2/luen/develop/workspace/doris/be/src/util/threadpool.cpp:48
22# doris::ThreadPool::dispatch_thread() at /mnt/disk2/luen/develop/workspace/doris/be/src/util/threadpool.cpp:533
23# void std::__invoke_impl<void, void (doris::ThreadPool::*&)(), doris::ThreadPool*&>(std::__invoke_memfun_deref, void (doris::ThreadPool::*&)(), doris::ThreadPool*&) at /mnt/disk2/luen/develop/tools/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:74
24# std::__invoke_result<void (doris::ThreadPool::*&)(), doris::ThreadPool*&>::type std::__invoke<void (doris::ThreadPool::*&)(), doris::ThreadPool*&>(void (doris::ThreadPool::*&)(), doris::ThreadPool*&) at /mnt/disk2/luen/develop/tools/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:96
25# void std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>::__call<void, , 0ul>(std::tuple<>&&, std::_Index_tuple<0ul>) at /mnt/disk2/luen/develop/tools/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:420
26# void std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>::operator()<, void>() at /mnt/disk2/luen/develop/tools/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:503
27# void std::__invoke_impl<void, std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>&>(std::__invoke_other, std::_Bind<void (doris::ThreadPool::*(doris::ThreadPool*))()>&) at /mnt/disk2/luen/develop/tools/ldb_toolchain/bin/../l
ib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:61

```

### What You Expected?

Normally.

### How to Reproduce?

1. create table 
```sql
 CREATE TABLE `wikipedia2` (
  `title` varchar(512) NOT NULL,
  `content` text NULL,
  `redirect` varchar(1024) NULL,
  `namespace` varchar(1024) NULL,
  INDEX idx_title (`title`) USING INVERTED PROPERTIES("parser" = "unicode", "support_phrase" = "true") COMMENT '',
  INDEX idx_content (`content`) USING INVERTED PROPERTIES("parser" = "unicode", "support_phrase" = "true") COMMENT ''
) ENGINE=OLAP
UNIQUE KEY(`title`)
COMMENT 'OLAP'
DISTRIBUTED BY HASH(`title`) BUCKETS 12
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"is_being_synced" = "false",
"storage_format" = "V2",
"compression" = "ZSTD",
"enable_unique_key_merge_on_write" = "true",
"light_schema_change" = "true",
"disable_auto_compaction" = "false",
"enable_single_replica_compaction" = "false"
);
```
2. Turn on index compaction
3. Load data 
    wikipedia dataset

### Anything Else?

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by qidaye**
In MOW table with inverted index, when the table has deleted docs, the doc_id in rowid_conversion_map will be INT32_MAX. index_compact in CLucene is not handling it correctly. It will cause doc lost in specific terms and the postings will be incorrect. The index compaction process will fail.

Here are the changes:
1. Remove INT32_MAX out of destPostingsQueues in CLucene to handle deleted doc right.
2. Add debug code and switch for index compaction in Doris
3. Add debug_index_compaction operation in index_tool

## Proposed changes

Issue Number: close #32084

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by qidaye** [comment]
run buildall

**Comment by qidaye** [comment]
run buildall

File: be/src/olap/compaction.cpp [hunk 1]
```diff
@@ -472,6 +474,45 @@ Status Compaction::do_compaction_impl(int64_t permits) {
                 dest_index_files[i] = prefix;
             }
 
+            // Only write info files when debug index compaction is enabled.
+            // The files are used to debug index compaction and works with index_tool.
+            if (config::debug_inverted_index_compaction) {
+                auto write_json_to_file = [&](const nlohmann::json& json_obj,
+                                              const std::string& file_name) {
+                    io::FileWriterPtr file_writer;
+                    std::string file_path = fmt::format("/tmp/{}.json", file_name);
```
Comment (airborne12) [inline be/src/olap/compaction.cpp:483]: use tmp dir in config?

**Review by zzzxl1993 — APPROVED** [review]
LGTM

File: be/src/olap/compaction.cpp (hunk 1 above)
Comment (qidaye) [reply be/src/olap/compaction.cpp]: Change it to sys_log_dir.

**Comment by qidaye** [comment]
run buildall

**Pull request merged**